README.md
# Deck.gl, React.js and Refugee Data

[![Build Status](https://travis-ci.org/hopetambala/dps-deck-refugee.svg?branch=master)](https://travis-ci.org/hopetambala/dps-deck-refugee)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/fccd6dec0e2a4488963b94722eee7a07)](https://www.codacy.com/manual/hopetambala/dps-deck-refugee?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=hopetambala/dps-deck-refugee&amp;utm_campaign=Badge_Grade)
[![codecov](https://codecov.io/gh/hopetambala/dps-deck-refugee/branch/master/graph/badge.svg)](https://codecov.io/gh/hopetambala/dps-deck-refugee)
![](https://img.shields.io/badge/react-✓-blue.svg)
![](https://img.shields.io/badge/deck-✓-blueviolet.svg)

## Demos
[Map](https://un-refugee-asylum-resettlement.herokuapp.com/)



## Description
<p align="middle">
    <img src="images/explain.png">
</p>

First Viz (Departures)
- Top ten countries of origin
- Country of Asylum in the Americas
- All Countries for resettlement


## Resources 

### Data
[UNHCR](https://www.unhcr.org/en-us/data.html)

## Libraries

## Project Layout

## Build and run

### Install
```
npm install
```
### Run
```
npm run start-local
```
## Tutorial & Development

## Extras



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.1.1 (2019-09-27)


src/App.js
import React, {Component} from 'react';
import './App.css';
import { StaticMap } from 'react-map-gl';
import { AmbientLight, PointLight, LightingEffect } from '@deck.gl/core';
import {ArcLayer} from '@deck.gl/layers';
import DeckGL from '@deck.gl/react';

import { ControlPanelComponent } from "./components/control-panel";

// Set your mapbox token here
const MAPBOX_TOKEN = "pk.eyJ1IjoiaHBiYWxhIiwiYSI6ImNrMXZyNWFscjB2N2szY3FmMHdodXZ2NjMifQ.PZQEuVD4WAHGTPd4yT5YFQ"; // eslint-disable-line

// Source data CSV
const DATA_URL ='https://raw.githubusercontent.com/hopetambala/dps-deck-refugee/master/data/2_RSQSubmissions_cleaned.json'; // eslint-disable-line

const INITIAL_VIEW_STATE = {
  //longitude: -70.1627,
  //latitude: 18.7357,
  latitude: -10.0,
  longitude: 0.0,
  zoom: 4,
  minZoom: 3,
  maxZoom: 20,
  pitch: 40.5,
  bearing: -27.396674584323023
};

const colorRange = [
  [1, 152, 189],
  [73, 227, 206],
  [216, 254, 181],
  [254, 237, 177],
  [254, 173, 84],
  [209, 55, 78]
];

const elevationScale = {min: 1, max: 50};

export const inFlowColors = [
  [255, 255, 204],
  [199, 233, 180],
  [127, 205, 187],
  [65, 182, 196],
  [29, 145, 192],
  [34, 94, 168],
  [12, 44, 132]
];

export const outFlowColors = [
  [255, 255, 178],
  [254, 217, 118],
  [254, 178, 76],
  [253, 141, 60],
  [252, 78, 42],
  [227, 26, 28],
  [177, 0, 38]
];


class App extends Component {
  static get defaultColorRange() {
    return colorRange;
  }

  constructor(props) {
    super(props);
    this.state = {
      data:null,
      elevationScale: elevationScale.max
    };
  }

  _renderLayers() {
    const data = this.state.data;
    const strokeWidth = 2;
    const strokeHeight = .5;
  
      return [
        new ArcLayer({
          id: 'arc',
          data: data,
          getSourcePosition: d => d["Country of Asylum Coordinates"],
          getTargetPosition: d => d["Country of Resettlement Coordinates"],
          // getSourceColor: d => (d.gain > 0 ? inFlowColors : outFlowColors)[d.quantile],
          getSourceColor: colorRange[0],
          // getTargetColor: d => (d.gain > 0 ? outFlowColors : inFlowColors)[d.quantile],
          getTargetColor: colorRange[4],
          getWidth: strokeWidth,
          getHeight: strokeHeight
        })
      ];
    }
  componentDidMount = () =>{
    this.fetchData()
  }
  
  async fetchData(){
    const response = await fetch(DATA_URL);
    const records = await response.json();
    
    console.log(records)

    this.setState({
      data:records
    })
  }


  render() {
    const {mapStyle = 'mapbox://styles/mapbox/dark-v9'} = this.props;

    return (
      <div>
        <DeckGL
          layers={this._renderLayers()}
          //effects={[lightingEffect]}
          initialViewState={INITIAL_VIEW_STATE}
          controller={true}
        >
          <StaticMap
            reuseMaps
            mapStyle={mapStyle}
            preventStyleDiffing={true}
            mapboxApiAccessToken={MAPBOX_TOKEN}
          />
        </DeckGL>
        <ControlPanelComponent />

      </div>

    );
  }
}

export default App;


src/components/control-panel.js
import React, {Component} from 'react';
import styled from 'styled-components';

const Div = styled.div`
    position: absolute;
    top: 0;
    width: 344px;
    background: #fff;
    box-shadow: 0 0 4px rgba(0,0,0,.15);
    margin: 24px;
    padding: 12px 24px;
    max-height: 96%;
    overflow-x: hidden;
    overflow-y: overlay;
    outline: none;
    right: 0;
`;

export class ControlPanelComponent extends Component {

    render(){
        return(
            <Div id="control-panel">

                <div>
                    <h3>United Nations Refugee Resettlement</h3>
                    <p>Asylum seeking to Resettlement from 2001 to 2019</p>
                    <p>The layer creates an arc from the country of orign to the asylum</p>
                    {/* <div><div>Accidents<b>140.1K</b></div></div> */}
                </div>
                {/* <div>
                    <label>Radius</label>
                    <input id="radius" type="range" min="1000" max="20000" step="1000" value="1000"></input>
                    <span id="radius-value"></span>
                </div>
                <div>
                    <label>Coverage</label>
                    <input id="coverage" type="range" min="0" max="1" step="0.1" value="1"></input>
                    <span id="coverage-value"></span>
                </div>
                <div>
                    <label>Upper Percentile</label>
                    <input id="upperPercentile" type="range" min="90" max="100" step="1" value="100"></input>
                    <span id="upperPercentile-value"></span>
                </div> */}
            </Div>
        )
    }

}



app.js
import React, {Component} from 'react';
import {render} from 'react-dom';
import {StaticMap} from 'react-map-gl';
import {PhongMaterial} from '@luma.gl/core';
import {AmbientLight, PointLight, LightingEffect} from '@deck.gl/core';
import {HexagonLayer} from '@deck.gl/aggregation-layers';
import DeckGL from '@deck.gl/react';

import { ControlPanelComponent } from "./control-panel";

// Set your mapbox token here
const MAPBOX_TOKEN = "pk.eyJ1IjoiaHBiYWxhIiwiYSI6ImNrMXZyNWFscjB2N2szY3FmMHdodXZ2NjMifQ.PZQEuVD4WAHGTPd4yT5YFQ"; // eslint-disable-line

// Source data CSV
const DATA_URL =
  'https://raw.githubusercontent.com/uber-common/deck.gl-data/master/examples/3d-heatmap/heatmap-data.csv'; // eslint-disable-line

const ambientLight = new AmbientLight({
  color: [255, 255, 255],
  intensity: 1.0
});

const pointLight1 = new PointLight({
  color: [255, 255, 255],
  intensity: 0.8,
  position: [-0.144528, 49.739968, 80000]
});

const pointLight2 = new PointLight({
  color: [255, 255, 255],
  intensity: 0.8,
  position: [-3.807751, 54.104682, 8000]
});

const lightingEffect = new LightingEffect({ambientLight, pointLight1, pointLight2});

const material = new PhongMaterial({
  ambient: 0.64,
  diffuse: 0.6,
  shininess: 32,
  specularColor: [51, 51, 51]
});

const INITIAL_VIEW_STATE = {
  longitude: -1.4157267858730052,
  latitude: 52.232395363869415,
  zoom: 6.6,
  minZoom: 5,
  maxZoom: 15,
  pitch: 40.5,
  bearing: -27.396674584323023
};

const colorRange = [
  [1, 152, 189],
  [73, 227, 206],
  [216, 254, 181],
  [254, 237, 177],
  [254, 173, 84],
  [209, 55, 78]
];

const elevationScale = {min: 1, max: 50};

/* eslint-disable react/no-deprecated */
export class App extends Component {
  static get defaultColorRange() {
    return colorRange;
  }

  constructor(props) {
    super(props);
    this.state = {
      elevationScale: elevationScale.min
    };
  }

  _renderLayers() {
    const {data, radius = 1000, upperPercentile = 100, coverage = 1} = this.props;

    return [
      new HexagonLayer({
        id: 'heatmap',
        colorRange,
        coverage,
        data,
        elevationRange: [0, 3000],
        elevationScale: data && data.length ? 50 : 0,
        extruded: true,
        getPosition: d => d,
        onHover: this.props.onHover,
        opacity: 1,
        pickable: Boolean(this.props.onHover),
        radius,
        upperPercentile,
        material,

        transitions: {
          elevationScale: 3000
        }
      })
    ];
  }

  render() {
    const {mapStyle = 'mapbox://styles/mapbox/dark-v9'} = this.props;

    return (
      <div>
        <DeckGL
          layers={this._renderLayers()}
          effects={[lightingEffect]}
          initialViewState={INITIAL_VIEW_STATE}
          controller={true}
        >
          <StaticMap
            reuseMaps
            mapStyle={mapStyle}
            preventStyleDiffing={true}
            mapboxApiAccessToken={MAPBOX_TOKEN}
          />
        </DeckGL>
        <ControlPanelComponent />

      </div>

    );
  }
}

export function renderToDOM(container) {
  render(<App />, container);

  require('d3-request').csv(DATA_URL, (error, response) => {
    if (!error) {
      const data = response.map(d => [Number(d.lng), Number(d.lat)]);
      render(<App data={data} />, container);
    }
  });
}
